controllers/apps/operations/restore.go

```diff
@@ -0,0 +1,202 @@
+// It will check the cluster status and update the OpsRequest status.
+// If the cluster is running, it will update the OpsRequest status to Complete.
+// If the cluster is failed, it will update the OpsRequest status to Failed.
+// If the cluster is not running, it will update the OpsRequest status to Running.
+func (r RestoreOpsHandler) ReconcileAction(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) (appsv1alpha1.OpsPhase, time.Duration, error) {
+	opsRequest := opsRes.OpsRequest
+	clusterDef := opsRequest.Spec.ClusterRef
+
+	// get cluster
+	cluster := &appsv1alpha1.Cluster{}
+	if err := cli.Get(reqCtx.Ctx, client.ObjectKey{
+		Namespace: opsRequest.GetNamespace(),
+		Name:      clusterDef,
+	}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			_ = PatchClusterNotFound(reqCtx.Ctx, cli, opsRes)
+		}
+		return appsv1alpha1.OpsFailedPhase, 0, err
+	}
+
+	// check if the cluster is running
+	if cluster.Status.Phase == appsv1alpha1.RunningClusterPhase {
+		return appsv1alpha1.OpsSucceedPhase, 0, nil
+	} else if cluster.Status.Phase == appsv1alpha1.FailedClusterPhase {
+		return appsv1alpha1.OpsFailedPhase, 0, fmt.Errorf("restore failed")
+	}
+	return appsv1alpha1.OpsRunningPhase, 0, nil
+}
+
+// SaveLastConfiguration saves last configuration to the OpsRequest.status.lastConfiguration
+func (r RestoreOpsHandler) SaveLastConfiguration(reqCtx intctrlutil.RequestCtx, cli client.Client, opsResource *OpsResource) error {
+	return nil
+}
+
+func restoreClusterFromBackup(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRequest *appsv1alpha1.OpsRequest, clusterDef string) (*appsv1alpha1.Cluster, error) {
+	backupName := opsRequest.Spec.RestoreSpec.BackupName
+	restoreTimeStr := opsRequest.Spec.RestoreSpec.RestoreTimeStr
+	volumeRestorePolicy := opsRequest.Spec.RestoreSpec.VolumeRestorePolicy
+
+	// check if the backup exists
+	backup := &dpv1alpha1.Backup{}
+	if err := cli.Get(reqCtx.Ctx, client.ObjectKey{
+		Name:      backupName,
+		Namespace: opsRequest.Namespace,
+	}, backup); err != nil {
+		return nil, err
+	}
+
+	// check if the backup whether is completed
+	if backup.Status.Phase != dpv1alpha1.BackupPhaseCompleted {
+		return nil, fmt.Errorf("backup %s is not completed", backupName)
+	}
+	if len(backup.Labels[constant.AppInstanceLabelKey]) == 0 {
+		return nil, fmt.Errorf(`missing source cluster in backup "%s", "app.kubernetes.io/instance" is empty in labels`, backupName)
+	}
+
+	restoreTimeStr, err := restore.FormatRestoreTimeAndValidate(opsRequest.Spec.RestoreSpec.RestoreTimeStr, backup)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the cluster object from backup
+	clusterObj, err := getClusterObjFromBackup(cli, backup)
+	if err != nil {
+		return nil, err
+	}
+	restoreAnnotation, err := restore.GetRestoreFromBackupAnnotation(backup, volumeRestorePolicy, len(clusterObj.Spec.ComponentSpecs), clusterObj.Spec.ComponentSpecs[0].Name, restoreTimeStr)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
better to place it in a function(getClusterObjFromBackup)  for building a cluster object and